stdlib/result.gr

```diff
@@ -0,0 +1,94 @@
+export let isOk = (v) => {
+    match (v) {
+        | Ok(_)  => true
+        | _      => false
+    }
+};
+
+export let isErr = (v) => !isOk(v);
+
+export let toOption = (v) => {
+    match (v) {
+        | Ok(x)  => Some(x)
+        | _      => None
+    }
+}
+
+export let flatMapOk = (v, fn) => {
+    match (v) {
+        | Ok(x)  => fn(x)
+        | _      => v
+    }
+}
+
+export let flatMapErr = (v, fn) => {
+    match (v) {
+        | Err(e) => fn(e)
+        | _      => v
+    }
+}
+
+export let flatMap = (v, fnOk, fnErr) => {
+    match (v) {
+        | Ok(x)  => fnOk(x)
+        | Err(x) => fnErr(x)
+    }
+}
```

Comment:
Can this fool the type system if I pass 2 methods that neither return a Result type? cc @ospencer

Reply:
I wouldn't call this fooling it.
The answer should be yes, since `fnOk` and `fnErr` should both be constrained to return the same type.
Either way, they need to return the same type, and it won't be a problem.

Reply:
That's the intended behavior, right? Since this is `flatMap` rather than `map`? The typechecker should just have to make sure `fnOk` and `fnErr` return the same type.

Reply:
Isn't `flatMap` usually expected to return the same type as the value? Like `bind` in the OCaml stdlib?

Reply:
Perhaps I mixed them up. Currently, `map` re-wraps the output automatically.

Reply:
You didn't mix them up, the implementation is correct, but the type system will cause (what should be) invalid usage to work.

```
Result.flatMap(Ok(1), (o) => "test", (e) => "test_err")
```

Should be invalid because `flatMap` functions should always return a Result type.

Reply:
If we don't want to wrap the output, I suppose this could just be solved with type annotations.

Reply:
I guess I don't really see the problem. Why is it necessary to arbitrarily constrain user's functions?

Reply:
I feel like it's fine. If we required you to return a result, it would just be the same as `map`. @phated were you thinking of this as a combination of `flatten` and `map`? I think that might be the confusion. I would expect a function that takes a `Result<Result<a, e> e>` and called `flatMap` to produce a `Result<b, f>`.

This could potentially be named `reduce` or something like that, and `flatMap` would do this ☝️

Reply:
Rust's `Result` `flatten`, for example: https://doc.rust-lang.org/beta/core/result/enum.Result.html#method.flatten

Reply:
Yeah, Rust calls this function `map_or_else`: https://doc.rust-lang.org/beta/core/result/enum.Result.html#method.map_or_else

Reply:
@ospencer no, our `flatMap` should be similar to `and_then` in Rust (which is also the same in Elm). We talked about that in my Option PR

This is the OCaml type for [`bind`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Result.html#VALbind):
```ocaml
val bind : ('a, 'e) result -> ('a -> ('b, 'e) result) -> ('b, 'e) result
```

And Rust's [`and_then`](https://doc.rust-lang.org/beta/core/result/enum.Result.html#method.and_then):
```rust
pub fn and_then<U, F: FnOnce(T) -> Result<U, E>>(self, op: F) -> Result<U, E>
```

And Belt's [`flatMap`](https://rescript-lang.org/apis/latest/belt/result#flatmap):
```reason
let flatMap: (t<'a, 'c>, 'a => t<'b, 'c>) => t<'b, 'c>
```

Reply:
Now that I'm looking at it, `flatMapOk` if the implementation that every other language use for their version of `flatMap`. Should we just rename that and delete this?

Reply:
The normal definitions, where `f` is the type wrapping the generics `a` and `b`, e.g. `Maybe(a)` or `Maybe(b)`:

- **`map`**: `f a -> (a -> b) -> f b` where `f` is the type wrapping the inner type
- **`flatMap`** (or `bind`)`: `f a -> (a -> f b) -> f b`

You'll see this exact same definition (with appropriate variation in syntax) across Haskell, Elm, OCaml, Rust, Swift, F♯… and this isn't arbitrary! It's because `map` and `flatMap` are well-defined interfaces (in category theory, the interfaces are named *functor* and *monad* respectively, and they really are just extremely-useful interfaces).

An example of how conflating these things gets weird: JavaScript's `Promise` chains and `await`, which are so difficult to type correctly that the TypeScript team is *still* working on better ways to plumb them through that type system, and it has more special cases than nearly any other part of the type system I can think of. That is so in part because `Promise` combines the `map` and `flatMap` interface—exactly the confusion 

Additionally, I'd argue that you *want* your core types to be well-defined type-wise (rather than relying on inference for them) because the inference for everything *else* needs to fall out correctly from these.

So for example with `Result`, `map` and `flatMap` would normally be constrained like this:

```ts
export let map = <t, u, e>(
  result: Result<t, e>,
  fn: (t: t) => u
): Result<u, e> => {
  match (result) {
    | Ok(v) => Ok(fn(v))
    | Err(e) => Err(e)
  }
}

export let flatMap = <t, u, e>(
  result: Result<t, e>,
  fn: (t: t) => Result<u, e>
): Result<u, e> => {
  match (result) {
    | Ok(v) => fn(v)
    | Err(e) => Err(e)
  }
}
```

This is because each of these defaults to one specific "path" through a set of variants for any given type with variants, e.g. `Option.map`/`Maybe.map` applies the function to the `Some`/`Just` case (for obvious reasons), `Result.map` to the `Ok` case, `Either` to the `Right` case, etc. Covering *both* scenarios with a function of that name is not only a bit unusual, it actually violates the interface! However, having the *additional* interfaces (e.g. as Rust supplies with names like `map_or_else`) is totally fine and *doesn't* violate those standard interfaces.

Reply:
@chriskrycho that's helpful and is in line with what @phated said.

Thus, I believe the API should look like this:

```
map : (a -> b, Result<a, e>) -> Result<b, e>
# formerly mapOk
# map(fn, r) is Ok(fn(a)) if r is Ok(a) and is r otherwise

mapErr : (e -> f, Result<a, e>) -> Result<a, f>
# mapErr(fn, r) is Err(fn(e)) if r is Err(e) and is r otherwise

mapWithDefault : (a -> b, b, Result<a, e>) -> b
# new function
# mapWithDefault(fn, default, r) is fn(a) if r is Ok(a) and is default otherwise

mapWithDefaultFn : (a -> b, e -> b, Result<a, e>) -> b
# formerly flatMap
# mapWithDefaultFn(fnOk, fnErr, r) is fnOk(a) if r is Ok(a) and fnErr(e) if r is Err(e)


flatMap : (a -> Result<b, e>, Result<a, e>) -> Result<b, e>
# formerly flatMapOk
# flatMap(fn, r) is fn(a) if r is Ok(a) and is r otherwise

flatMapErr : (e -> Result<a, f>, Result<a, e>) -> Result<a, f>
# flatMapErr(fn, r) is fn(e) if r is Err(e) and is r otherwise

# There is no flatMapWithDefault or flatMapWithDefaultFn, as these are both easily accomplished with mapWithDefault and mapWithDefaultFn, but it would not hurt to provide them.
```

This makes our API match that of OCaml/Rust/Elm/Belt & friends, while using (what I believe to be) the most clear names.

One thing to note: I've swapped the argument orders to put the data last. 🙂 

@c0d3d, I know that's quite a bit to ask for; if you'd like, I can make those changes.

Reply:
Looks perfect and I’m always a big fan of data-last!